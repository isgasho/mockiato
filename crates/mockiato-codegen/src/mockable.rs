use crate::generate::{generate_mock, GenerateMockOptions};
use crate::parse::mockable_attr::MockableAttr;
use crate::parse::trait_decl::TraitDecl;
use crate::spanned::SpannedUnstable;
use crate::Error;
use proc_macro::{Diagnostic, Level, Span, TokenStream};
use quote::quote;
use syn::{AttributeArgs, Item, ItemTrait};

#[derive(Default)]
pub(crate) struct Mockable;

impl Mockable {
    pub(crate) fn new() -> Self {
        Self::default()
    }

    pub(crate) fn expand(&self, attr: AttributeArgs, item: Item) -> TokenStream {
        let item_backup = item.clone();

        #[doc(hidden)]
        macro try_or_return($expr: expr) {
            match $expr {
                Ok(value) => value,
                Err(_) => return TokenStream::from(quote! { #item_backup }),
            }
        }

        let mockable_attr = try_or_return!(MockableAttr::parse(attr).map_err(Error::emit));
        let item_trait = try_or_return!(extract_item_trait(item).map_err(Error::emit));
        let trait_decl = try_or_return!(TraitDecl::parse(item_trait.clone())
            .map_err(|err| err
                .emit_with(|d| d.span_note(Span::call_site(), "Required for mockable traits"))));

        let generated_mock = generate_mock(
            &trait_decl,
            GenerateMockOptions {
                custom_struct_ident: mockable_attr.name_attr.map(|attr| attr.ident),
                force_static_lifetimes: mockable_attr.static_attr.is_some(),
            },
        );

        TokenStream::from(quote! {
            #item_trait
            #generated_mock
        })
    }
}

fn extract_item_trait(item: Item) -> Result<ItemTrait, Error> {
    match item {
        Item::Trait(item_trait) => Ok(item_trait),
        _ => Err(Error::Diagnostic(
            Diagnostic::spanned(
                item.span_unstable(),
                Level::Error,
                "Only traits can be made mockable",
            )
            .span_note(Span::call_site(), "Required because of this attribute"),
        )),
    }
}
