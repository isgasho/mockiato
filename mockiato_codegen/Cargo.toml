[package]
name = "mockiato_codegen"
version = "0.1.0"
authors = [
    "Ruben Schmidmeister <ruben@myelin.ch>",
    "Jeremy Stucki <jeremy@myelin.ch>",
]
edition = "2018"
license = "MIT"

[lib]
proc-macro = true

[dependencies]
quote = "0.6"
heck = "0.3"

[dependencies.syn]
version = "0.15"
features = ["full", "extra-traits", "visit-mut"]

[dependencies.proc-macro2]
version = "0.4"
features = ["nightly"]
