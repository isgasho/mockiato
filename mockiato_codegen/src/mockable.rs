use std::sync::RwLock;
use syntax::ast::{self, Ident, VariantData, DUMMY_NODE_ID};
use syntax::ext::base::{Annotatable, ExtCtxt, MultiItemDecorator};
use syntax::ext::build::AstBuilder;
use syntax::ptr::P;
use syntax_pos::Span;

use crate::parse::mockable_attr::MockableAttr;
use crate::parse::name_attr::NameAttr;
use crate::parse::trait_bounds::TraitBounds;
use crate::parse::trait_decl::TraitDecl;
use crate::definition_id::DefId;
use crate::trait_bound_resolver::{TraitBoundResolver, TraitBoundType};

pub(crate) struct Mockable {
    trait_bound_resolver: RwLock<Box<dyn TraitBoundResolver>>,
}

const TRAIT_BOUND_RESOLVER_ERR: &str = "Internal Error: Trait Bound Resolver is poisoned";

impl Mockable {
    pub(crate) fn new(trait_bound_resolver: Box<dyn TraitBoundResolver>) -> Self {
        Self {
            trait_bound_resolver: RwLock::new(trait_bound_resolver),
        }
    }

    fn register_current_trait(&self, trait_bound_def_id: DefId, trait_decl: &TraitDecl) {
        self.trait_bound_resolver
            .write()
            .expect(TRAIT_BOUND_RESOLVER_ERR)
            .register_mocked_trait(trait_bound_def_id, &trait_decl);
    }

    fn mock_trait_bound_impls(&self, cx: &mut ExtCtxt, trait_decl: &TraitDecl) {
        let trait_bounds = TraitBounds::parse(trait_decl).0;
        for trait_bound in trait_bounds {
            let identifier = trait_bound.identifier;
            let trait_bound_resolver = self
                .trait_bound_resolver
                .read()
                .expect(TRAIT_BOUND_RESOLVER_ERR);
            let trait_bound_type = trait_bound_resolver
                .resolve_trait_bound(&identifier);
            self.mock_trait_bound_type(cx, trait_bound.span, &trait_bound_type);
        }
    }

    fn mock_trait_bound_type(&self, cx: &mut ExtCtxt, sp: Span, trait_bound_type: &Option<TraitBoundType>) {
        match *trait_bound_type {
            None => {
                cx.parse_sess
                .span_diagnostic
                .mut_span_err(sp, "The referenced trait has has not been marked as #[mockable] or doesn't exist")
                .help("Mockable traits are handled from top to bottom, try declaring this trait earlier.")
                .emit();
            },
            _ => {}
        }
    }
}

impl MultiItemDecorator for Mockable {
    fn expand(
        &self,
        cx: &mut ExtCtxt,
        _sp: Span,
        meta_item: &ast::MetaItem,
        item: &Annotatable,
        push: &mut dyn FnMut(Annotatable),
    ) {
        let trait_decl = match TraitDecl::parse(cx, item) {
            Ok(trait_decl) => trait_decl,
            Err(_) => return,
        };

        let mockable_attr = match MockableAttr::parse(cx, meta_item) {
            Some(mockable_attr) => mockable_attr,
            None => return,
        };

        // self.register_current_trait(ruben_give_id, &trait_decl);
        
        let mock_struct_ident = mock_struct_ident(&trait_decl, mockable_attr.name_attr);

        let mut mock_struct = cx
            .item_struct(
                meta_item.span,
                mock_struct_ident,
                VariantData::Unit(DUMMY_NODE_ID),
            ).into_inner();

        if let Some(derive_attr) = mockable_attr.derive_attr {
            mock_struct.attrs.push(derive_attr.expand(cx));
        }

        push(Annotatable::Item(P(mock_struct)));

        // self.mock_trait_bound_impls(&mut cx, &trait_decl);
    }
}

fn mock_struct_ident(trait_decl: &TraitDecl, name_attr: Option<NameAttr>) -> Ident {
    name_attr
        .map(|attr| attr.expand())
        .unwrap_or_else(|| Ident::from_str(&format!("{}Mock", trait_decl.ident)))
}
