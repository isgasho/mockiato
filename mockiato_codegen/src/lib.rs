#![crate_type = "dylib"]
#![feature(
    quote,
    concat_idents,
    plugin_registrar,
    rustc_private,
    decl_macro,
    custom_attribute,
    underscore_imports,
)]

use rustc_plugin::Registry;
use syntax::ext::base::SyntaxExtension;
use syntax::symbol::Symbol;

mod context;
mod definition_id;
mod derive_resolver;
mod mockable;
mod parse;
mod trait_bound_resolver;

use self::mockable::Mockable;
use self::trait_bound_resolver::TraitBoundResolverImpl;

#[plugin_registrar]
pub fn plugin_registrar(reg: &mut Registry) {
    reg.register_syntax_extension(
        Symbol::intern("mockable"),
        SyntaxExtension::MultiDecorator(Box::new(Mockable::new(Box::new(
            TraitBoundResolverImpl::new(),
        )))),
    );
}
