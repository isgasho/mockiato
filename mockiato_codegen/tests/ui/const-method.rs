use mockiato_codegen::mockable;

#[mockable]
trait Foo {
    const fn bar(&self, baz: u64);
}

fn main() {}
