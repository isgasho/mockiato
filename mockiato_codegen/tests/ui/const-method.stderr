error: `const` methods are not supported
 --> $DIR/const-method.rs:5:5
  |
5 |     const fn bar(&self, baz: u64);
  |     ^^^^^
  |
note: Required because of #[mockable] on the trait declaration
 --> $DIR/const-method.rs:4:1
  |
4 | trait Foo {
  | ^^^^^

error: aborting due to previous error

